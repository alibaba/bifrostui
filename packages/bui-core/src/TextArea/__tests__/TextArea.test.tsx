import React from 'react';
import { fireEvent, isConformant, render, screen, userEvent } from 'testing';
import { TextArea } from '..';

describe('TextArea', () => {
  const rootClass = 'bui-textarea';

  isConformant({
    className: rootClass,
    displayName: 'BuiTextArea',
    Component: TextArea,
    skip: [
      'component-has-root-ref',
      'component-handles-classNames',
      'component-handles-style',
    ],
  });

  it('should render with value', () => {
    render(<TextArea value="123" />);
    const textarea = screen.getByRole('textbox');
    expect(textarea).toHaveValue('123');
  });

  it('should controlled by value', () => {
    const handleChange = jest.fn((e, data) => data.value);
    render(<TextArea placeholder="textarea" onChange={handleChange} />);
    const textarea = screen.getByRole('textbox');
    userEvent.type(textarea, '1234');
    expect(handleChange).toReturnWith('1234');
    expect(textarea).toHaveValue('1234');
  });

  it('should render with defaultValue', () => {
    render(<TextArea defaultValue="234" />);
    const textarea = screen.getByRole('textbox');
    expect(textarea).toHaveValue('234');
  });

  it('should render placeholder', () => {
    render(<TextArea placeholder="place input content" />);
    const textarea = screen.getByRole('textbox');
    expect((textarea as HTMLTextAreaElement).placeholder).toBe(
      'place input content',
    );
  });

  it('should render by rows', () => {
    render(<TextArea rows={3} />);
    const textarea = screen.getByRole('textbox');
    expect((textarea as HTMLTextAreaElement).rows).toBe(3);
  });

  it('should have max length', () => {
    render(<TextArea maxLength={10} />);
    const textarea = screen.getByRole('textbox');
    expect((textarea as HTMLTextAreaElement).maxLength).toBe(10);
  });

  it('can auto size', () => {
    render(
      <TextArea
        autoSize
        value="test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize."
      />,
    );
    const textarea = screen.getByRole('textbox');
    expect(textarea).toMatchSnapshot();
  });

  it('can auto size', () => {
    render(<TextArea autoSize={{ minHeight: 50, maxHeight: 100 }} />);
    const textarea = screen.getByRole('textbox');

    Object.defineProperty(textarea, 'scrollHeight', {
      value: 30,
      configurable: true,
    });
    fireEvent.change(textarea, { target: { value: '1' } });
    expect(textarea.style.height).toBe(`30px`);

    Object.defineProperty(textarea, 'scrollHeight', {
      value: 200,
    });
    fireEvent.change(textarea, {
      target: {
        value:
          'test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.',
      },
    });
    expect(textarea.style.height).toBe(`100px`);
  });

  it('should auto focus', () => {
    render(<TextArea autoFocus />);
    const textarea = screen.getByRole('textbox');
    expect(document.activeElement).toBe(textarea);
  });

  it('should show input count', () => {
    const { container } = render(<TextArea value="test" showCount />);
    const count = container.querySelector(`.${rootClass}-count`);
    expect(count).toBeInTheDocument();
    expect(count).toHaveTextContent('4');
  });

  it('trigger on change', () => {
    const fakeChange = jest.fn((e, v) => v.value);
    render(<TextArea onChange={fakeChange} />);
    const textarea = screen.getByRole('textbox');
    userEvent.type(textarea, '1234');
    expect(fakeChange).toReturnWith('1234');
  });

  it('trigger focus', () => {
    render(<TextArea value="test" />);

    const textarea = screen.getByRole('textbox');
    textarea.focus();
    expect(textarea).toHaveFocus();
  });

  it('trigger blur', () => {
    render(<TextArea value="test" />);

    const textarea = screen.getByRole('textbox');
    textarea.focus();
    expect(textarea).toHaveFocus();
    textarea.blur();
    expect(textarea).not.toHaveFocus();
  });
});
