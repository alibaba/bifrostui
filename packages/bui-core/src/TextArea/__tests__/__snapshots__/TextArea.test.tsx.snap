// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea can auto size 1`] = `
<textarea
  class="bui-textarea-content"
  maxlength="-1"
  rows="2"
  style="height: 0px;"
>
  test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.test autoSize.
</textarea>
`;
